index.js

```diff
@@ -72,62 +72,17 @@ const bindConnection = (options, aedesHandler, conn, req = {}) => {
 }
 
 const extractConnectionDetails = (options, aedesHandler, conn, req = {}) => {
-  const rewindableStream = conn.pipe(new Rewindable())
-
-  conn.on('data', (buffer) => {
+  const onData = (buffer) => {
     const protocol = options.protocolDecoder(conn, buffer, req)
-
-    let stream = null
-    if (buffer && protocol.data) {
-      const removeProxyHeader = removeFromStream(
-        buffer.length - protocol.data.length
-      )
-      stream = rewindableStream.rewind().pipe(removeProxyHeader)
-    } else {
-      stream = rewindableStream.rewind()
-    }
     req.connDetails = protocol
-    aedesHandler(stream, req)
-  })
-}
-
-class Rewindable extends Transform {
-  constructor () {
-    super()
-    this.accumulator = []
-  }
-
-  _transform (chunk, encoding, callback) {
-    this.accumulator.push(chunk)
-    callback()
-  }
-
-  rewind () {
-    const stream = new PassThrough()
-    this.accumulator.forEach((chunk) => stream.write(chunk))
-    return stream
+    conn.removeListener('data', onData)
+    conn.pause()
+    conn.unshift(protocol.data || buffer)
+    aedesHandler(conn, req)
```

Comment:
This should be better than the previous stream chain

Reply:
@mcollina

Reply:
Yes! Note that `buffer` _might_ not contain the full protocol info, so you need to do some work to handle that.

Reply:
@mcollina When you said full protocol you were thinking of MQTT ? If you think of Proxy header, it seems safe to assume that the whole header will be present in the buffer.